src/debugger/gdb-stub.c

```diff
@@ -433,7 +448,62 @@ static void _processQSupportedCommand(struct GDBStub* stub, const char* message)
 		}
 		message = end + 1;
 	}
-	strncpy(stub->outgoing, "swbreak+;hwbreak+", GDB_STUB_MAX_LINE - 4);
+	strncpy(stub->outgoing, "swbreak+;hwbreak+;qXfer:features:read+;qXfer:memory-map:read+", GDB_STUB_MAX_LINE - 4);
+}
+
+static void _processQXferCommand(struct GDBStub* stub, const char* params, const char* data) {
+	unsigned offset = 0;
+	unsigned length = 0;
+
+	unsigned index = 0;
+	for (index = 0; params[index] != ','; ++index) {
+		offset <<= 4;
+		offset |= _hex2int(&params[index], 1);
+	}
+
+	++index;
+	unsigned int paramsLength = strlen(params);
+	for (; index + 3 < paramsLength; ++index) {
+		length <<= 4;
+		length |= _hex2int(&params[index], 1);
+	}
+
+	if (length + 1 > GDB_STUB_MAX_LINE) {
+		length = GDB_STUB_MAX_LINE - 1;
+	}
+
+	memset(stub->outgoing, 0, GDB_STUB_MAX_LINE);
+	if (strlen(data) < length + offset) {
+		length = strlen(data) - offset;
+		stub->outgoing[0] = 'l';
+	} else {
+		stub->outgoing[0] = 'm';
+	}
+	strlcpy(&stub->outgoing[1], &data[offset], length + 1);
```

Comment:
The + 1 being "needed" shows that you're calculating the length wrong.

Reply:
I debugged and I am actually not sure: GDB doesn't seem to count the null terminator as part of its requested length. So what strlcpy does is say GDB requests 651 bytes (which it does on my end), this will actually put 650 bytes then put the null terminator which makes it 651 length. I checked with a debugger and despite the fact I pass in a size of 651 (which is what GDB asked me), I end up with stub->outgoing being length 651 with a null terminator, but remember that the first byte doesn't count because it's m or l so I end up sending one byte short.

Am I correct? because the length is definitely calculated correctly at 651 when I get to this line, but this somehow copies 650 bytes instead of 651.

Reply:
The discrepancy is the null terminator byte. `strlcpy` is supposed to take the size of the _buffer_ to ensure that after copying it is properly null terminated. If gdb requests a number of bytes and you use that directly then that means you're calculating the...number of bytes it's asking for, not the length of the buffer.